### ISSUE TITLE
Docker build stopped working

### ISSUE THREAD

**Issue by jmacedoit** [facebook/duckling#472]
I'm getting an error when building a docker image.
On a fresh clone, I'm doing:
```
docker build --no-cache -t i-duckling .
``` 
and I get:
```
Sending build context to Docker daemon  18.67MB
Step 1/15 : FROM haskell:8 AS builder
 ---> 095b182d9077
Step 2/15 : RUN apt-get update -qq &&   apt-get install -qq -y libpcre3 libpcre3-dev build-essential --fix-missing --no-install-recommends &&   apt-get clean &&   rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
 ---> Running in e38d353cc538
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package libpcrecpp0v5:amd64.
(Reading database ... 16696 files and directories currently installed.)
Preparing to unpack .../0-libpcrecpp0v5_2%3a8.39-3_amd64.deb ...
Unpacking libpcrecpp0v5:amd64 (2:8.39-3) ...
Selecting previously unselected package bzip2.
Preparing to unpack .../1-bzip2_1.0.6-8.1_amd64.deb ...
Unpacking bzip2 (1.0.6-8.1) ...
Selecting previously unselected package patch.
Preparing to unpack .../2-patch_2.7.5-1+deb9u2_amd64.deb ...
Unpacking patch (2.7.5-1+deb9u2) ...
Selecting previously unselected package dpkg-dev.
Preparing to unpack .../3-dpkg-dev_1.18.25_all.deb ...
Unpacking dpkg-dev (1.18.25) ...
Selecting previously unselected package build-essential.
Preparing to unpack .../4-build-essential_12.3_amd64.deb ...
Unpacking build-essential (12.3) ...
Selecting previously unselected package libpcre16-3:amd64.
Preparing to unpack .../5-libpcre16-3_2%3a8.39-3_amd64.deb ...
Unpacking libpcre16-3:amd64 (2:8.39-3) ...
Selecting previously unselected package libpcre32-3:amd64.
Preparing to unpack .../6-libpcre32-3_2%3a8.39-3_amd64.deb ...
Unpacking libpcre32-3:amd64 (2:8.39-3) ...
Selecting previously unselected package libpcre3-dev:amd64.
Preparing to unpack .../7-libpcre3-dev_2%3a8.39-3_amd64.deb ...
Unpacking libpcre3-dev:amd64 (2:8.39-3) ...
Setting up bzip2 (1.0.6-8.1) ...
Setting up patch (2.7.5-1+deb9u2) ...
Processing triggers for libc-bin (2.24-11+deb9u4) ...
Setting up libpcrecpp0v5:amd64 (2:8.39-3) ...
Setting up libpcre32-3:amd64 (2:8.39-3) ...
Setting up libpcre16-3:amd64 (2:8.39-3) ...
Setting up libpcre3-dev:amd64 (2:8.39-3) ...
Setting up dpkg-dev (1.18.25) ...
Setting up build-essential (12.3) ...
Processing triggers for libc-bin (2.24-11+deb9u4) ...
Removing intermediate container e38d353cc538
 ---> e31e73153ebf
Step 3/15 : RUN mkdir /log
 ---> Running in 9f773ae720ce
Removing intermediate container 9f773ae720ce
 ---> 111234831577
Step 4/15 : WORKDIR /duckling
 ---> Running in f3e351a94abc
Removing intermediate container f3e351a94abc
 ---> 4d6e1e6565bd
Step 5/15 : ADD stack.yaml .
 ---> d7ba4f3d6a56
Step 6/15 : ENV LANG=C.UTF-8
 ---> Running in c732a9bd7a5f
Removing intermediate container c732a9bd7a5f
 ---> d3884b819646
Step 7/15 : RUN stack setup
 ---> Running in 5c7dfc8bbf4a
Stack looks for packages in the directories configured in
the 'packages' and 'extra-deps' fields defined in your stack.yaml
The current entry points to /duckling/,
but no .cabal or package.yaml file could be found there.
The command '/bin/sh -c stack setup' returned a non-zero code: 1
```

Switching the order between`RUN stack setup` and `ADD . .` appears to fix it.
The image size remains at 135Mb.

@patapizza @pranas

**Comment by pranas** [comment]
I think some changes were made to master after my work on Dockerfile. The ordering is meant to optimize caching. We have `ADD stack.yaml .` before `stack setup` so dependencies are only rebuilt when `stack.yaml` changes. Looking at the error message I think we either need to rename `stack.yaml` to `package.yaml` or add `duckling.cabal` as well. I will experiment with this later today.

**Comment by pranas** [comment]
Turns out, the cause was an update to haskell. The build is using `haskell:8` as base image which now refers to haskell 8.8. I think the best approach would be to lock the base image to `haskell:8.6.3` which matches the version being used by Travis CI.

@patapizza what do you think? should I open up a PR for this?

**Comment by jmacedoit** [comment]
@pranas Being new to the Haskell ecosystem , there are some things that still confuse me. In this case: 
- shouldn't stack install its own standalone version of ghci when necessary (therefore rendering the base image haskell version irrelevant (unless the error has something to do with the stack version itself))
- why don't we use a base image that supplies a ghci version that matches the one needed by the stack resolver in `stack.yml`

As I said, I'm new to the Haskell ecosystem so if this comment doesn't make sense, my apologies in advance 😄

**Comment by pranas** [comment]
@jmacedoit I'm very new to Haskell ecosystem too but

> shouldn't stack install its own standalone version of ghci when necessary (therefore rendering the base image haskell version irrelevant (unless the error has something to do with the stack version itself))

I think it does, but the error was related to stack version being too new and no longer supporting the current config in this repo.

> why don't we use a base image that supplies a ghci version that matches the one needed by the stack resolver in stack.yml? Therefore stack wouldn't need to download ghci when building.

Good point, I think that means we should lock to `8.0.2`.

**Comment by jmacedoit** [comment]
@pranas 

> Good point, I think that means we should lock to 8.0.2.

Or, or or - we upgrade to 8.4.4 (or greater). I found that this version still works without additional modifications, builds on OSx (yes, there is this bug - https://github.com/facebook/duckling/issues/446 - duckling doesn't build on OSx at the moment, and people are deleting their stack.yml)  and as a bonus we get an haskell-ide-engine supported version. We solve 3 existing problems (Docker, build on OSx and bad tooling support).

Only thing I still can't run are tests...But can't say it is because of 8.4.4 since I can't build with 8.0.2either way...

**Comment by pranas** [comment]
Well, the project is already being tested with 8.6.3 on Travis CI (although it might just start on that version and download the 8.0 in the process). However I'm not in a position to decide anything. @jmacedoit You will have to wait for the feedback from someone more involved in the project.

**Comment by pheel** [comment]
I did not know about the caching optimization, but moving the `ADD . .` line up is one of the things [this PR](https://github.com/facebook/duckling/pull/499) does. Also updated the runtime image from debian:stretch since the binary from build step is no longer compatible with it.

**Comment by jayaddison** [comment]
@jmacedoit @pranas I too am not super familiar with the state of Haskell tooling, but have attempted a fix for this build issue in PR #518.  If you get a chance to test that out and provide feedback / suggestions, I'd be glad to incorporate them.  It's a small diff and I'd be happy to learn if there's a more correct fix!

**Comment by chessai** [comment]
fix landed on master. feel free to re-open if there are any issues.
